runtime/coverage_test.go

```diff
@@ -28,8 +28,280 @@ import (
+
+	t.Parallel()
+
+	// Represents line numbers with statements.
+	// We see that line 9 appears twice, meaning that it has 2 statements.
+	// For example: "if let index = self.index(s, until, startIndex) {"
+	// However, we cannot distinguish hits between the two statements, for the time being.
+	// TBD
+	statementLines := []int{3, 4, 5, 7, 9, 9, 11}
+	locationCoverage := NewLocationCoverage(statementLines)
+
+	assert.Equal(
+		t,
+		map[int]int{3: 0, 4: 0, 5: 0, 7: 0, 9: 0, 11: 0},
+		locationCoverage.LineHits,
+	)
+	assert.ElementsMatch(
+		t,
+		[]int{3, 4, 5, 7, 9, 11},
+		locationCoverage.MissedLines(),
+	)
+	assert.Equal(t, 6, locationCoverage.Statements)
+	assert.Equal(t, "0.0%", locationCoverage.Percentage)
+	assert.Equal(t, 0, locationCoverage.CoveredLines())
+}
+
+func TestLocationCoverageAddLineHit(t *testing.T) {
+
+	t.Parallel()
+
+	statementLines := []int{3, 4, 5, 7, 9, 9, 11}
+	locationCoverage := NewLocationCoverage(statementLines)
+
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(7)
+	locationCoverage.AddLineHit(9)
+	// Line 15 was not included in the statementLines array,
+	// however we want it to be tracked. This will help
+	// find out about cases where the inspector does not
+	// find all the statements. We should also discuss
+	// if the Statements counter should be increased in
+	// this case.
+	// TBD
+	locationCoverage.AddLineHit(15)
+
+	assert.Equal(
+		t,
+		map[int]int{3: 2, 4: 0, 5: 0, 7: 1, 9: 1, 11: 0, 15: 1},
+		locationCoverage.LineHits,
+	)
+	assert.Equal(t, 6, locationCoverage.Statements)
+	assert.Equal(t, "66.7%", locationCoverage.Percentage)
+}
+
+func TestLocationCoverageCoveredLines(t *testing.T) {
+
+	t.Parallel()
+
+	statementLines := []int{3, 4, 5, 7, 9, 9, 11}
+	locationCoverage := NewLocationCoverage(statementLines)
+
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(7)
+	locationCoverage.AddLineHit(9)
+	locationCoverage.AddLineHit(15)
+
+	assert.Equal(t, 4, locationCoverage.CoveredLines())
+}
+
+func TestLocationCoverageMissedLines(t *testing.T) {
+	t.Parallel()
+
+	statementLines := []int{3, 4, 5, 7, 9, 9, 11}
+	locationCoverage := NewLocationCoverage(statementLines)
+
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(7)
+	locationCoverage.AddLineHit(9)
+	locationCoverage.AddLineHit(15)
+
+	assert.ElementsMatch(
+		t,
+		[]int{4, 5, 11},
+		locationCoverage.MissedLines(),
+	)
+}
+
+func TestLocationCoveragePercentage(t *testing.T) {
+
+	t.Parallel()
+
+	statementLines := []int{3, 4, 5}
+	locationCoverage := NewLocationCoverage(statementLines)
+
+	locationCoverage.AddLineHit(3)
+	locationCoverage.AddLineHit(4)
+	locationCoverage.AddLineHit(5)
+	locationCoverage.AddLineHit(15)
+
+	// One side effect of counting line hits that are not included
+	// in the statementLines array, is a percentage > 100.
+	// TBD
+	assert.Equal(t, "133.3%", locationCoverage.Percentage)
```

Comment:
Maybe just saturate at 100%?

Reply:
That makes sense! Fixed in https://github.com/onflow/cadence/pull/2296/commits/4bc05a38b54b77a7b76ff78c3db9968856f62051